api/workload/v2/stream.go

```diff
@@ -0,0 +1,177 @@
+package workload
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/api/workload"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const _unixPathPrefix = "unix://"
+
+// streamManager manages connection streams
+type streamManager struct {
+	// StreamChan is a channel of streams for fetching X509 SVIDs. It is updated whenever a new stream is created.
+	StreamChan chan *managedStream
+	// ConnectionChan is a stream of connection events.
+	ConnectionChan chan bool
+	ctx            context.Context
+	logger         *logrus.Logger
+	addr           string
+	reconnectChan  chan struct{}
+	forceStart     bool
+}
+
+type managedStream struct {
+	client workload.SpiffeWorkloadAPI_FetchX509SVIDClient
+	closer io.Closer
+}
+
+// Recv blocks for an SVID response and is context-aware.
+func (s *managedStream) Recv(ctx context.Context) (res *workload.X509SVIDResponse, err error) {
+	done := make(chan struct{})
+	go func() {
+		res, err = s.client.Recv()
+		close(done)
+	}()
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-done:
+		return res, err
+	}
+}
+
+// Close closes the stream and the underlying connection.
+func (s *managedStream) Close() error {
+	var errs []string
+	if err := s.client.CloseSend(); err != nil {
+		errs = append(errs, err.Error())
+	}
+	if err := s.closer.Close(); err != nil {
+		errs = append(errs, err.Error())
+	}
+	return errors.New(strings.Join(errs, "; "))
+}
+
+func newStreamManager(ctx context.Context, logger *logrus.Logger, addr string, forceStart bool) (*streamManager, error) {
+	if !strings.HasPrefix(addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", addr)
+	}
+	return &streamManager{
+		StreamChan:     make(chan *managedStream, 1),
+		ctx:            ctx,
+		logger:         logger,
+		addr:           addr,
+		reconnectChan:  make(chan struct{}, 1),
+		ConnectionChan: make(chan bool, 1),
+		forceStart:     forceStart,
+	}, nil
+}
+
+// Reconect informs the stream manager that the current stream is unusable.
+func (c *streamManager) Reconnect() {
+	c.reconnectChan <- struct{}{}
+}
+
+// Stop stops the stream manager
+func (c *streamManager) Stop() {
+	c.logger.Debug("Shutting down stream manager.")
+	if c.StreamChan != nil {
+		close(c.StreamChan)
+	}
+	if c.ConnectionChan != nil {
+		close(c.ConnectionChan)
+		c.logger.WithField("queued", len(c.ConnectionChan)).Debug("Emptying connection chan.")
+		for range c.ConnectionChan {
+		}
+		c.ConnectionChan = nil
+	}
+}
+
+// Start starts the stream manager.
+//
+// This blocks until a stream is established and will return an error if a stream
+// can't be established within the context's deadline.
+func (c *streamManager) Start(ctx context.Context) error {
+	stream, closer, err := c.newStream(ctx, c.addr)
```

Comment:
From my read we need to create a subcontext here to allow `Recv` with a context to correctly close the stream without leaking it.

```
// NewStream creates a new Stream for the client side. This is typically
// called by generated code. ctx is used for the lifetime of the stream.
//
// To ensure resources are not leaked due to the stream returned, one of the following
// actions must be performed:
//
//      1. Call Close on the ClientConn.
//      2. Cancel the context provided.
//      3. Call RecvMsg until a non-nil error is returned. A protobuf-generated
//         client-streaming RPC, for instance, might use the helper function
//         CloseAndRecv (note that CloseSend does not Recv, therefore is not
//         guaranteed to release all resources).
//      4. Receive a non-nil, non-io.EOF error from Header or SendMsg.
//
// If none of the above happen, a goroutine and a context will be leaked, and grpc
// will not call the optionally-configured stats handler with a stats.End message.
```

Reply:
heh i missed this in the docs apparently.. probably using a subcontext would be simpler. right now i'm carrying the ClientConn around as part of the *managedStream and so it occurs as part of the client close

Reply:
Which means that we potentially pile up those goroutines until a full client shutdown right?

Reply:
we close a prior stream before opening a new one, but let me add some tests for this